src/websocket_server/websocket_server.c

```diff
@@ -0,0 +1,164 @@
+/**
+ * @file websocket_server.c  Websocket Server Demo
+ *
+ * The demo can for instance be tested by pasting this into
+ * a web browser javascript console:
+ *	var ws = new WebSocket('ws://localhost:3456');
+ *	ws.addEventListener('open', () => ws.send('hello'));
+ *	ws.addEventListener('close', (e) => console.log('close', e));
+ *	ws.addEventListener('message', (e) => console.log('msg', e));
+ *
+ * Copyright (C) 2018 Creytiv.com
+ */
+
+#include <re.h>
+
+
+struct websocket {
+	struct le le;				/* list entry */
+	struct websock_conn *conn;	/* Websocket connection */
+};
+
+static struct websock *server;
+
+/* Linked list of active Websocket connections */
+static struct list connl;
+
+/* called upon reception of  SIGINT, SIGALRM or SIGTERM */
+static void signal_handler(int sig)
+{
+	re_printf("terminating on signal %d...\n", sig);
+
+	/* destroy active Websocket connections */
+	list_flush(&connl);
+
+	websock_shutdown(server);
+}
+
+static void websock_shutdown_handler(void *arg)
+{
+	(void)arg;
+	re_printf("websocket server shutting down\n");
+
+	/* stop libre main loop */
+	re_cancel();
+}
+
+/* destructor called when reference count on conn reach zero */
+static void destructor(void *arg)
+{
+	struct websocket *ws = arg;
+
+	/* remove this connection entry from the list */
+	list_unlink(&ws->le);
+
+	/* clean up local Websocket connection state */
+	mem_deref(ws->conn);
+}
+
+/* called when a message is received from the other side */
+static void srv_websock_recv_handler(const struct websock_hdr *hdr,
+					 struct mbuf *mb, void *arg)
+{
+	struct websocket *ws = arg;
+	int err;
+
+	err = websock_send(ws->conn, hdr->opcode, "%b",
+		mbuf_buf(mb), mbuf_get_left(mb));
+	if (err)
+		re_fprintf(stderr, "ws send error: %m\n", err);
+}
+
+/* called when the websocket is closed by the other side */
+static void srv_websock_close_handler(int err, void *arg)
+{
+	struct websocket *ws = arg;
+
+	re_fprintf(stderr, "ws close %m\n", err);
+	mem_deref(ws);
+}
+
+/* HTTP request handler */
+static void http_req_handler(struct http_conn *conn,
+				 const struct http_msg *msg, void *arg)
+{
+	struct websocket *ws;
+	int err;
+
+	/* allocate connection state */
+	ws = mem_zalloc(sizeof(*ws), destructor);
+	if (!ws) {
+		err = ENOMEM;
+		re_fprintf(stderr, "http req handler alloc error: %m\n", err);
+		return;
+	}
+
+	err = websock_accept(&ws->conn, server, conn, msg,
+				 0, srv_websock_recv_handler,
+				 srv_websock_close_handler, ws);
+	if (err) {
+		re_fprintf(stderr, "websocket accept error: %mßn", err);
```

Comment:
Some strange character seems to have sneaked in here, instead of '\' for linefeed escape.